internal/service/odb/cloud_exadata_infrastructure_data_source.go

```diff
@@ -0,0 +1,425 @@
+					"based on the physical core of a processor with hyper-threading enabled.",
+			},
+			"created_at": schema.StringAttribute{
+				Computed:    true,
+				Description: "The time when the Exadata infrastructure was created.",
+			},
+			"database_server_type": schema.StringAttribute{
+				Computed:    true,
+				Description: "The database server model type of the Exadata infrastructure. For the list of valid model names, use the ListDbSystemShapes operation.",
+			},
+			"storage_server_type": schema.StringAttribute{
+				Computed:    true,
+				Description: "The storage server model type of the Exadata infrastructure. For the list of valid model names, use the ListDbSystemShapes operation.",
+			},
+			names.AttrTags: tftags.TagsAttributeComputedOnly(),
+			"maintenance_window": schema.ObjectAttribute{
+				Computed:    true,
+				CustomType:  fwtypes.NewObjectTypeOf[cloudExadataInfraMaintenanceWindowDataSourceModel](ctx),
+				Description: "The maintenance window for the Exadata infrastructure.",
+				AttributeTypes: map[string]attr.Type{
+					"custom_action_timeout_in_mins": types.Int32Type,
+					"days_of_week": types.SetType{
+						ElemType: fwtypes.StringEnumType[odbtypes.DayOfWeekName](),
+					},
+					"hours_of_day": types.SetType{
+						ElemType: types.Int32Type,
+					},
+					"is_custom_action_timeout_enabled": types.BoolType,
+					"lead_time_in_weeks":               types.Int32Type,
+					"months": types.SetType{
+						ElemType: fwtypes.StringEnumType[odbtypes.MonthName](),
+					},
+					"patching_mode": fwtypes.StringEnumType[odbtypes.PatchingModeType](),
+					"preference":    fwtypes.StringEnumType[odbtypes.PreferenceType](),
+					"weeks_of_month": types.SetType{
+						ElemType: types.Int32Type,
+					},
+				},
+			},
+		},
+		Blocks: map[string]schema.Block{
+			"customer_contacts_to_send_to_oci": schema.SetNestedBlock{
+				Description: "Customer contact emails to send to OCI.",
+				CustomType:  fwtypes.NewSetNestedObjectTypeOf[customerContactDataSourceModel](ctx),
+				NestedObject: schema.NestedBlockObject{
+					Attributes: map[string]schema.Attribute{
+						"email": schema.StringAttribute{
+							Computed: true,
+						},
+					},
+				},
+			},
+		},
+	}
+}
+
+func (d *dataSourceCloudExadataInfrastructure) Read(ctx context.Context, req datasource.ReadRequest, resp *datasource.ReadResponse) {
+	conn := d.Meta().ODBClient(ctx)
+
+	var data cloudExadataInfrastructureDataSourceModel
+	resp.Diagnostics.Append(req.Config.Get(ctx, &data)...)
+	if resp.Diagnostics.HasError() {
+		return
+	}
+
+	out, err := FindOdbExaDataInfraForDataSourceByID(ctx, conn, data.CloudExadataInfrastructureId.ValueString())
+	if err != nil {
+		resp.Diagnostics.AddError(
+			create.ProblemStandardMessage(names.ODB, create.ErrActionReading, DSNameCloudExadataInfrastructure, data.CloudExadataInfrastructureId.String(), err),
+			err.Error(),
+		)
+		return
+	}
+	tagsRead, err := listTags(ctx, conn, *out.CloudExadataInfrastructureArn)
```

Comment:
you can set this data source to find tags by adding `// @Tags(identifierAttribute="arn")` as one of the tags. This will eliminate the call to `listTags` in this handler and be handled transparently.